runtime/src/main/scala/akka/grpc/internal/AkkaHttpClientUtils.scala

```diff
@@ -0,0 +1,226 @@
+        val src =
+          invokeWithMetadata(Source.single(request), descriptor.getFullMethodName, headers, descriptor, false, options)
+        val result = src.toMat(Sink.headOption)(Keep.right).run()
+        result.flatMap {
+          case Some(r) => Future.successful(r)
+          case None    => Future.failed(throw new IllegalStateException("Successful status code but no data found"))
+        }
+      }
+
+      override def invokeWithMetadata[I, O](
+          request: I,
+          headers: MetadataImpl,
+          descriptor: MethodDescriptor[I, O],
+          options: CallOptions): Future[GrpcSingleResponse[O]] = ???
+
+      override def invokeWithMetadata[I, O](
+          source: Source[I, NotUsed],
+          fqMethodName: String,
+          headers: MetadataImpl,
+          descriptor: MethodDescriptor[I, O],
+          streamingResponse: Boolean,
+          options: CallOptions): Source[O, Future[GrpcResponseMetadata]] = {
+        implicit val serializer: ProtobufSerializer[I] = descriptor
+        val deserializer: ProtobufSerializer[O] = descriptor
+        val httpRequest = GrpcRequestHelpers(
+          Uri(s"https://${settings.overrideAuthority.getOrElse(host)}/" + descriptor.getFullMethodName),
+          source)
+        Source.lazyFutureSource[O, Future[GrpcResponseMetadata]](() => {
+          singleRequest(httpRequest).map { response =>
+            {
+              Codecs.detect(response) match {
+                case Success(codec) =>
+                  log.info(s"XXX response $response started")
+                  implicit val reader: GrpcProtocolReader = GrpcProtocolNative.newReader(codec)
+                  val trailerPromise = Promise[immutable.Seq[HttpHeader]]()
+                  // Completed with success or failure based on grpc-status and grpc-message trailing headers
+                  val completionFuture: Future[Unit] = trailerPromise.future.flatMap(parseResponseStatus)
+                  completionFuture.foreach(_ => log.info(s"XXX response $response completion"))
+
+                  response.entity match {
+                    case Chunked(_, chunks) =>
+                      chunks
+                        .concat(Source
+                          .maybe[ChunkStreamPart]
+                          .mapMaterializedValue(promise => promise.completeWith(completionFuture.map(_ => None))))
+                        .map {
+                          case Chunk(data, _) =>
+                            data
+                          case LastChunk(_, trailer) =>
+                            trailerPromise.success(trailer)
+                            ByteString.empty
+                        }
+                        .watchTermination()((_, done) =>
+                          done.onComplete(c => {
+                            log.info(s"XXX response $response termination ${c.isSuccess}")
+                            trailerPromise.trySuccess(immutable.Seq.empty)
```

Comment:
Should the trailer promise fail if the stream failed?

Reply:
I'm not sure I understand the relation between `trailerPromise` and `completionFuture` with the whole stream.

I think it's unexpected complexity to capture the `trailers` value so it's available in the `GrpcResponseMetadata` below but I wonder if I'm missing something else.

Reply:
> I'm not sure I understand the relation between `trailerPromise` and `completionFuture` with the whole stream.

`completionFuture` makes sure the stream fails with the correct `io.grpc.StatusRuntimeException` when the trailer contains a non-successful status code.

> I think it's unexpected complexity to capture the `trailers` value so it's available in the `GrpcResponseMetadata` below but I wonder if I'm missing something else.

I agree it's unexpected complexity in the implementation, but I think it produces simplicity in the API.

In the 'regular' API, a failed streaming response is represented as the stream failing with a `io.grpc.StatusRuntimeException`. When using the 'power' API, you get access to the complete `akka.grpc.GrpcResponseMetadata` which contains all the detailed (leading and (asynchronously) trailing) header information.